document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresUnnestQueryTransformer.java

```diff
@@ -0,0 +1,267 @@
+      return unnestExpression;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class FilterIdentifierProvider implements SelectTypeExpressionVisitor {
+
+    @Override
+    public List<String> visit(AggregateExpression expression) {
+      return expression.getExpression().accept(this);
+    }
+
+    @Override
+    public List<String> visit(ConstantExpression expression) {
+      return List.of();
+    }
+
+    @Override
+    public List<String> visit(FunctionExpression expression) {
+      return expression.getOperands().stream()
+          .map(exp -> exp.<List<String>>accept(this))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<String> visit(IdentifierExpression expression) {
+      return List.of(expression.getName());
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class AndFiltersCollector implements FilterTypeExpressionVisitor {
+
+    @Override
+    public List<FilterTypeExpression> visit(LogicalExpression expression) {
+      if (expression.getOperator().equals(LogicalOperator.AND)) {
+        return expression.getOperands().stream()
+            .map(
+                filterTypeExpression ->
+                    filterTypeExpression.<List<FilterTypeExpression>>accept(this))
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+      }
+      return List.of(expression);
+    }
+
+    @Override
+    public List<FilterTypeExpression> visit(RelationalExpression expression) {
+      return List.of(expression);
+    }
+
+    @Override
+    public List<FilterTypeExpression> visit(KeyExpression expression) {
+      return List.of(expression);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class FilterToUnnestExpressionMatcher implements FilterTypeExpressionVisitor {
+    private final UnnestExpression unnestExpression;
+    private final FilterIdentifierProvider filterIdentifierProvider;
+
+    FilterToUnnestExpressionMatcher(
+        FilterIdentifierProvider filterIdentifierProvider, UnnestExpression unnestExpression) {
+      this.unnestExpression = unnestExpression;
+      this.filterIdentifierProvider = filterIdentifierProvider;
+    }
+
+    @Override
+    public Boolean visit(LogicalExpression expression) {
+      if (expression.getOperator().equals(LogicalOperator.OR)) {
+        return expression.getOperands().stream()
+            .filter(filterTypeExpression -> filterTypeExpression.accept(this))
+            .findFirst()
+            .isPresent();
+      }
+      return false;
+    }
+
+    @Override
+    public Boolean visit(RelationalExpression expression) {
+      List<String> lhs = expression.getLhs().accept(filterIdentifierProvider);
+      return lhs.stream()
+          .anyMatch(p -> p.contains(unnestExpression.getIdentifierExpression().getName()));
+    }
+
+    @Override
+    public Boolean visit(KeyExpression expression) {
```

Comment:
Any reason for using a wrapper Boolean type here (instead of the primitive one)?

Reply:
Parameterized required Object.